backpack/core/derivatives/batchnorm_nd.py

```diff
@@ -0,0 +1,312 @@
+        mat: Tensor,
+    ) -> Tensor:
+        _n_axis = self._get_n_axis(module)
+        out = mat.unsqueeze(1)
+        for _ in range(_n_axis):
+            out = out.unsqueeze(-1)
+        dim_expand: List[int] = [-1, module.input0.shape[0], -1]
+        for n in range(_n_axis):
+            dim_expand.append(module.input0.shape[2 + n])
+        return out.expand(*dim_expand)
+
+    def _bias_jac_t_mat_prod(
+        self,
+        module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d],
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+        mat: Tensor,
+        sum_batch: bool = True,
+    ) -> Tensor:
+        _n_dim: int = self._get_n_axis(module)
+        axis_sum: Tuple[int] = {
+            0: (1,) if sum_batch else None,
+            1: (1, 3) if sum_batch else (3,),
+            2: (1, 3, 4) if sum_batch else (3, 4),
+            3: (1, 3, 4, 5) if sum_batch else (3, 4, 5),
+        }[_n_dim]
+        return mat if axis_sum is None else mat.sum(dim=axis_sum)
+
+
+class BatchNorm1dDerivatives(BatchNormNdDerivatives):
+    """Derivatives for BatchNorm1d."""
+
+    def __init__(self):
+        """Initialization."""
+        super().__init__(n_dim=1)
+
+    def _residual_mat_prod(
+        self,
+        module: BatchNorm1d,
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+        mat: Tensor,
+    ) -> Tensor:
+        """Multiply with BatchNorm1d residual-matrix.
+
+        Paul Fischer (GitHub: @paulkogni) contributed this code during a research
+        project in winter 2019.
+
+        Details are described in
+
+        - `TODO: Add tech report title`
+          <TODO: Wait for tech report upload>_
+          by Paul Fischer, 2020.
+
+        Args:
+            module: module
+            g_inp: input gradients
+            g_out: output gradients
+            mat: matrix to multiply
+
+        Returns:
+            product
+        """
+        N = module.input0.size(0)
+        x_hat, var = self._get_normalized_input_and_var(module)
+        gamma = module.weight
+        eps = module.eps
+
+        factor = gamma / (N * (var + eps))
+
+        sum_127 = einsum("nc,vnc->vc", x_hat, mat)
+        sum_24 = einsum("nc->c", g_out[0])
+        sum_3 = einsum("nc,vnc->vc", g_out[0], mat)
+        sum_46 = einsum("vnc->vc", mat)
+        sum_567 = einsum("nc,nc->c", x_hat, g_out[0])
+
+        r_mat = -einsum("nc,vc->vnc", g_out[0], sum_127)
+        r_mat += (1.0 / N) * einsum("c,vc->vc", sum_24, sum_127).unsqueeze(1).expand(
+            -1, N, -1
+        )
+        r_mat -= einsum("nc,vc->vnc", x_hat, sum_3)
+        r_mat += (1.0 / N) * einsum("nc,c,vc->vnc", x_hat, sum_24, sum_46)
+
+        r_mat -= einsum("vnc,c->vnc", mat, sum_567)
+        r_mat += (1.0 / N) * einsum("c,vc->vc", sum_567, sum_46).unsqueeze(1).expand(
+            -1, N, -1
+        )
+        r_mat += (3.0 / N) * einsum("nc,vc,c->vnc", x_hat, sum_127, sum_567)
+
+        return einsum("c,vnc->vnc", factor, r_mat)
```

Comment:
This is some old code, that I just copied and cleaned a little bit.
Keep it?
Does it need a checker for training mode?
Does it generalize to more dimensions? (It seems to be written for 0d: inputs of shape (N, C). Therefore, it doesn't support BatchNorm1d: inputs of shape (N, C) and inputs of shape (N, C, L).)
Is the tech report out?

Reply:
This function is used by the [`HMP`](https://docs.backpack.pt/en/master/extensions.html#backpack.extensions.HMP) extension, and should be kept. It can be generalized to higher-order batch norms at a later stage, probably only in case of a feature request.

This function assumes train mode is on. A check for that would be good.

I PMed you regarding the tech report.

Reply:
Added checker.
Added tech report title.
Moved it to the parent class -> will throw `NotImplementedError` in most cases, but works for 0d